File: apps/backend/schemas/api_key.py
"""
Pydantic schemas for API Key management — Phase 5 PR #16.

Public shapes:
  - APIKeyCreateIn        — request body for POST /v1/api-keys
  - APIKeyCreateOut       — response with the plaintext (returned ONCE)
  - APIKeyListItem        — list-row shape (no plaintext, no hash)
  - APIKeyListPage        — paginated list wrapper
  - APIKeyScope           — Literal type alias for the closed scope set

Design notes:
  - APIKeyCreateOut.raw_key is the plaintext bearer string
    (``tos_<prefix>_<secret>``). It is returned exactly once at issuance and
    intentionally NOT stored on the server side. The list endpoint NEVER echoes
    the plaintext — clients must capture it from the create response or rotate
    the key.
  - APIKeyListItem omits ``key_hash`` so a leaky serializer (e.g. a future bug
    that round-trips ORM rows directly) cannot accidentally surface the hash.
  - Literal types on ``scope`` give us crisp OpenAPI + Pydantic v2 validation;
    a bogus value fails fast with a 422 RFC 7807 envelope.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# Closed scope set. ``Literal`` is friendlier to OpenAPI than a Postgres ENUM
# round-trip — the API accepts only these three values.
APIKeyScope = Literal["org", "team", "project"]


class APIKeyCreateIn(BaseModel):
    """Request body for creating a new API key.

    The CHECK constraint on the DB enforces scope coherence; we mirror it in
    Pydantic for fast client-side feedback. The router pre-validates that the
    actor can actually issue at the requested scope (super_admin → org,
    team_admin → team, team member → project).
    """

    name: str = Field(..., min_length=1, max_length=100)
    scope: APIKeyScope
    team_id: UUID | None = None
    project_id: UUID | None = None


class APIKeyCreateOut(BaseModel):
    """Response from POST /v1/api-keys.

    ``raw_key`` is the only place the plaintext is ever surfaced. The client
    is responsible for storing it (e.g. as a CI secret); a subsequent GET on
    the key returns the metadata-only :class:`APIKeyListItem` shape.
    """

    id: UUID
    key_prefix: str
    name: str
    scope: APIKeyScope
    team_id: UUID | None
    project_id: UUID | None
    created_by_user_id: UUID | None
    created_at: datetime
    raw_key: str = Field(
        ...,
        description=(
            "The plaintext bearer key (format: tos_<prefix>_<secret>). "
            "Returned exactly once at issuance; capture it client-side. "
            "Subsequent reads only return metadata."
        ),
    )


class APIKeyListItem(BaseModel):
    """List-row shape — never includes the plaintext or the hash."""

    id: UUID
    key_prefix: str
    name: str
    scope: APIKeyScope
    team_id: UUID | None
    project_id: UUID | None
    created_by_user_id: UUID | None
    created_at: datetime
    last_used_at: datetime | None
    revoked_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class APIKeyListPage(BaseModel):
    """Paginated list of API keys."""

    items: list[APIKeyListItem]
    total: int
    page: int
    page_size: int


__all__ = [
    "APIKeyCreateIn",
    "APIKeyCreateOut",
    "APIKeyListItem",
    "APIKeyListPage",
    "APIKeyScope",
]

File: apps/frontend/src/types/apiKey.ts
/**
 * API Key wire types — chore C (`/integrations` UI).
 *
 * Mirrors apps/backend/schemas/api_key.py. Snake_case is preserved for the
 * wire shape so the JSON round-trip is verbatim; the React layer uses
 * these types directly without renaming because every other typed wrapper
 * in @/lib/projectsApi follows the same convention.
 */

export type APIKeyScope = "org" | "team" | "project";

/** Response from POST /v1/api-keys — `raw_key` is shown ONCE. */
export interface APIKeyCreateOut {
  id: string;
  key_prefix: string;
  name: string;
  scope: APIKeyScope;
  team_id: string | null;
  project_id: string | null;
  created_by_user_id: string | null;
  created_at: string;
  /**
   * Plaintext bearer key (`tos_<prefix>_<secret>`). The backend returns this
   * exactly once at issuance; the SPA shows it in a "copy now" dialog and
   * never persists it. Subsequent reads return :class:`APIKeyListItem`.
   */
  raw_key: string;
}

/** Row shape returned by GET /v1/api-keys. NEVER includes the plaintext. */
export interface APIKeyListItem {
  id: string;
  key_prefix: string;
  name: string;
  scope: APIKeyScope;
  team_id: string | null;
  project_id: string | null;
  created_by_user_id: string | null;
  created_at: string;
  last_used_at: string | null;
  revoked_at: string | null;
}

export interface APIKeyListPage {
  items: APIKeyListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface APIKeyCreatePayload {
  name: string;
  scope: APIKeyScope;
  team_id?: string | null;
  project_id?: string | null;
}

export interface ListAPIKeysParams {
  scope?: APIKeyScope;
  team_id?: string;
  project_id?: string;
  include_revoked?: boolean;
  page?: number;
  page_size?: number;
}
